In [5]:
import pandas as pd
import numpy as np

In [6]:
df=pd.read_csv('../data/raw/WA_Fn-UseC_-HR-Employee-Attrition.csv')

In [24]:
df.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

In [25]:
df.isnull().sum().sum()

np.int64(0)

In [26]:
df.duplicated().sum()

np.int64(0)

In [34]:
df.columns[df.nunique().sort_values(ascending=True)==1]

Index(['Age', 'Attrition', 'BusinessTravel'], dtype='object')

In [35]:
df['EmployeeNumber'].nunique()

1470

In [36]:
df.shape

(1470, 35)

In [37]:
df['Attrition'].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

In [38]:
df['Department'].value_counts()

Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64

In [39]:
df['Gender'].value_counts()

Gender
Male      882
Female    588
Name: count, dtype: int64

In [40]:
df['OverTime'].value_counts()

OverTime
No     1054
Yes     416
Name: count, dtype: int64

In [43]:
categorical_columns = df.select_dtypes(include=['object']).columns
for col in categorical_columns:
    print("\n",col)
    print(df[col].unique())


 Attrition
['Yes' 'No']

 BusinessTravel
['Travel_Rarely' 'Travel_Frequently' 'Non-Travel']

 Department
['Sales' 'Research & Development' 'Human Resources']

 EducationField
['Life Sciences' 'Other' 'Medical' 'Marketing' 'Technical Degree'
 'Human Resources']

 Gender
['Female' 'Male']

 JobRole
['Sales Executive' 'Research Scientist' 'Laboratory Technician'
 'Manufacturing Director' 'Healthcare Representative' 'Manager'
 'Sales Representative' 'Research Director' 'Human Resources']

 MaritalStatus
['Single' 'Married' 'Divorced']

 Over18
['Y']

 OverTime
['Yes' 'No']


In [45]:
numerical_columns=df.select_dtypes(include=['int64']).columns
for col in numerical_columns:
    print("\n",col)
    print(df[col].max(),df[col].min())



 Age
60 18

 DailyRate
1499 102

 DistanceFromHome
29 1

 Education
5 1

 EmployeeCount
1 1

 EmployeeNumber
2068 1

 EnvironmentSatisfaction
4 1

 HourlyRate
100 30

 JobInvolvement
4 1

 JobLevel
5 1

 JobSatisfaction
4 1

 MonthlyIncome
19999 1009

 MonthlyRate
26999 2094

 NumCompaniesWorked
9 0

 PercentSalaryHike
25 11

 PerformanceRating
4 3

 RelationshipSatisfaction
4 1

 StandardHours
80 80

 StockOptionLevel
3 0

 TotalWorkingYears
40 0

 TrainingTimesLastYear
6 0

 WorkLifeBalance
4 1

 YearsAtCompany
40 0

 YearsInCurrentRole
18 0

 YearsSinceLastPromotion
15 0

 YearsWithCurrManager
17 0


In [47]:
(df['YearsInCurrentRole']>df['YearsAtCompany']).sum()


np.int64(0)

In [50]:
(df['YearsWithCurrManager']>df['YearsAtCompany']).sum()


np.int64(0)

In [72]:
l_col=df.columns[df.nunique()==1]

In [80]:
l=['EmployeeNumber']
for i in l_col:
    l.append(i)
df_clean=df.drop(columns=l)

In [81]:
df_clean.shape

(1470, 31)

In [82]:
df_clean.columns[df_clean.nunique()==1]

Index([], dtype='object')

In [83]:
df_clean.to_csv('../data/cleaned/cleaned_hr_emp_data.csv',index=False)

# Data Cleaning Summary

## Data Cleaning Actions

- No meesing values are found in datasets.
- No duplicate rows were found.
- Datatypes were reviewed and found to be appropriate for analysis.
- constant columns with only one unique value were identified:
    - `EmployeeCount`
    - `Over18`
    - `StandardHours`
- `EmployeeNumber` was a column was identified which is not useful for analysis the Attrition because all are unique.
These Four columns are removed from the dataset.


## Final Dataset

The Cleaned dataset Contains:
- **1470 rows**
- **31 columns**
The Dataset is now ready for Exploratory Data Analysis (EDA) to investiagte the factors that associated with employee attrition.